filebeat/input/filestream/internal/input-logfile/store.go

```diff
@@ -149,6 +159,86 @@ func openStore(log *logp.Logger, statestore StateStore, prefix string) (*store,
 	}, nil
 }
 
+func newSourceStore(s *store, identifier *sourceIdentifier) *sourceStore {
+	return &sourceStore{
+		store:      s,
+		identifier: identifier,
+	}
+}
+
+func (s *sourceStore) FindCursorMeta(src Source, v interface{}) error {
+	key := s.identifier.ID(src)
+	return s.store.findCursorMeta(key, v)
+}
+
+func (s *sourceStore) UpdateMetadata(src Source, v interface{}) error {
+	key := s.identifier.ID(src)
+	return s.store.updateMetadata(key, v)
+}
+
+func (s *sourceStore) Remove(src Source) error {
+	key := s.identifier.ID(src)
+	return s.store.remove(key)
+}
+
+// CleanIf sets the TTL of a resource if the predicate return true.
+func (s *sourceStore) CleanIf(pred func(v Value) bool) {
+	if !s.identifier.configuredUserID {
+		return
+	}
+
+	s.store.ephemeralStore.mu.Lock()
+	defer s.store.ephemeralStore.mu.Unlock()
+
+	for key, res := range s.store.ephemeralStore.table {
+		if !s.identifier.MatchesInput(key) {
+			continue
+		}
+
+		if !res.lock.TryLock() {
+			continue
+		}
+
+		remove := pred(res)
+		if remove {
+			s.store.UpdateTTL(res, 0)
+		}
+		res.lock.Unlock()
+	}
+}
+
+// UpdateIdentifiers copies an existing resource to a new ID and marks the previous one
+// for removal.
+func (s *sourceStore) UpdateIdentifiers(getNewID func(v Value) (string, interface{})) {
+	s.store.ephemeralStore.mu.Lock()
+	defer s.store.ephemeralStore.mu.Unlock()
+
+	for key, res := range s.store.ephemeralStore.table {
+		if !s.identifier.MatchesInput(key) {
+			continue
+		}
+
+		if !res.lock.TryLock() {
+			continue
+		}
+
+		newKey, updatedMeta := getNewID(res)
+		if len(newKey) > 0 && res.internalState.TTL > 0 {
+			if _, ok := s.store.ephemeralStore.table[newKey]; ok {
+				res.lock.Unlock()
+				continue
+			}
+
+			r := res.copyWithNewKey(newKey)
+			r.cursorMeta = updatedMeta
+			r.stored = false
+			s.store.writeState(r)
```

Comment:
Note (maybe document): pending updates due to events that have not yet been ACKed will not be included in the copy. Collection on the copy will start from the last known ACKed position. This can lead to duplicates if configurations are adapted for inputs with the same ID are changed.